ksqldb-engine/src/main/java/io/confluent/ksql/services/SandboxedSchemaRegistryClient.java

```diff
@@ -36,21 +43,213 @@
 
   static SchemaRegistryClient createProxy(final SchemaRegistryClient delegate) {
     Objects.requireNonNull(delegate, "delegate");
-
-    return LimitedProxyBuilder.forClass(SchemaRegistryClient.class)
-        .swallow("register", anyParams(), 123)
-        .swallow("getId", anyParams(), 123)
-        .forward("getAllSubjects", methodParams(), delegate)
-        .forward("getSchemaById", methodParams(int.class), delegate)
-        .forward("getLatestSchemaMetadata", methodParams(String.class), delegate)
-        .forward("getSchemaBySubjectAndId", methodParams(String.class, int.class), delegate)
-        .forward("testCompatibility",
-            methodParams(String.class, ParsedSchema.class), delegate)
-        .swallow("deleteSubject", methodParams(String.class), Collections.emptyList())
-        .forward("getVersion", methodParams(String.class, ParsedSchema.class), delegate)
-        .build();
+    return new SandboxSchemaRegistryCache(delegate);
   }
 
   private SandboxedSchemaRegistryClient() {
   }
+
+  static final class SandboxSchemaRegistryCache implements SchemaRegistryClient {
+    private final MockSchemaRegistryClient mockedClient = new MockSchemaRegistryClient();
+    private final SchemaRegistryClient delegate;
+
+    private SandboxSchemaRegistryCache(final SchemaRegistryClient delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    public Optional<ParsedSchema> parseSchema(
+        final String schemaType,
+        final String schemaString,
+        final List<SchemaReference> references) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public int register(final String subject, final ParsedSchema parsedSchema)
+        throws RestClientException, IOException {
+      return mockedClient.register(subject, parsedSchema);
+    }
+
+    @Override
+    public int register(
+        final String subject,
+        final ParsedSchema parsedSchema,
+        final int version,
+        final int id) {
+      return -1; // swallow
+    }
+
+    @Deprecated
+    @Override
+    public Schema getById(final int id) {
+      throw new UnsupportedOperationException();
+
+    }
+
+    @Override
+    public ParsedSchema getSchemaById(final int id) throws RestClientException, IOException {
+      try {
+        return delegate.getSchemaById(id);
+      } catch (RestClientException e) {
+        // if we don't find the schema in SR, we try to get it from the sandbox cache
+        if (e.getStatus() == HttpStatus.SC_NOT_FOUND) {
+          return mockedClient.getSchemaById(id);
```

Comment:
The earlier version of this PR only returned from this line if the schema was present in the sandbox cache, and otherwise threw on the line below. This latest version of the PR now always returns from the mocked client representing the sandbox cache. Does the mocked client throw an error if the schema is not present? Otherwise there is a behavior change here (which might be fine, wondering if it's intentional).

Same comment/question for the other methods below too (`getSchemaBySubjectAndId`, `getLatestSchemaMetadata`, `getVersion`, and `getId`).

Reply:
It also throws if not found: https://github.com/confluentinc/schema-registry/blob/master/client/src/main/java/io/confluent/kafka/schemaregistry/client/MockSchemaRegistryClient.java#L123-L143.